### PR TITLE
Shutdown clusters when large number of nodes

### LINKED ISSUE #17386: Cluster failed to shutdown when # of nodes is large
Repo, start a cluster with more than 1000 nodes.
```
(anyscale) yic@ip-172-31-38-60:~/ray-w2/python (debug-info) $ ray down ~/check-config.yaml
Loaded cached provider configuration
If you experience issues with the cloud provider, try re-running the command with --no-config-cache.
Destroying cluster. Confirm [y/N]: y
Fetched IP: 54.202.153.35
Warning: Permanently added '54.202.153.35' (ECDSA) to the list of known hosts.
Did not find any active Ray processes.
Shared connection to 54.202.153.35 closed.
1050
Traceback (most recent call last):
  File "/mnt/yic/.local/anaconda3/envs/anyscale/bin/ray", line 33, in <module>
    sys.exit(load_entry_point('ray', 'console_scripts', 'ray')())
  File "/mnt/yic/ray-w2/python/ray/scripts/scripts.py", line 1944, in main
    return cli()
  File "/mnt/yic/.local/anaconda3/envs/anyscale/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/mnt/yic/.local/anaconda3/envs/anyscale/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/mnt/yic/.local/anaconda3/envs/anyscale/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/mnt/yic/.local/anaconda3/envs/anyscale/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/mnt/yic/.local/anaconda3/envs/anyscale/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/mnt/yic/ray-w2/python/ray/scripts/scripts.py", line 996, in down
    keep_min_workers)
  File "/mnt/yic/ray-w2/python/ray/autoscaler/_private/commands.py", line 450, in teardown_cluster
    provider.terminate_nodes(A)
  File "/mnt/yic/ray-w2/python/ray/autoscaler/_private/aws/node_provider.py", line 473, in terminate_nodes
    self.ec2.meta.client.terminate_instances(InstanceIds=node_ids)
  File "/mnt/yic/.local/anaconda3/envs/anyscale/lib/python3.7/site-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/mnt/yic/.local/anaconda3/envs/anyscale/lib/python3.7/site-packages/botocore/client.py", line 676, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (ResourceCountExceeded) when calling the TerminateInstances operation: You have exceeded the number of resources allowed in a single call of this type
```

### PR DESCRIPTION
**Pull request by hngenc**
## Why are these changes needed?

Ray currently fails to shutdown clusters with over 1000 nodes, because the AWS instance termination API only accept a maximum of 1000 instances to terminate in a single request.

To fix this, we loop over all the instances we want to terminate, breaking them into chunks of 1000.

## Related issue number

Closes #17386

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [X] This PR is not tested :(

I am not sure how to add tests for this. Should I add a release test that spins up and terminates 1001 instances? Or would that be considered too expensive?

---
This change is [Reviewable](https://reviewable.io/reviews/ray-project/ray/17642)

### CODE REVIEW

**Review by mwtian — APPROVED** [review]

**Review by AmeerHajAli — COMMENTED** [review]
@hngenc , is this mock testable?

**Comment by hngenc** [comment]
@AmeerHajAli There doesn't seem to be a mock test for `terminate_nodes` right now. Only `terminate_node` (without the `s`) is mocked, and that's just for `NodeProvider`s in general, rather than the AWS NodeProvider in particular:

https://github.com/ray-project/ray/blob/57b190c98770c0582d525076c261a104fc72a66a/python/ray/tests/test_autoscaler.py#L245

I'll try to create a new mock test in `python/ray/tests/aws/test_autoscaler_aws.py` for the `AWSNodeProvider`'s `terminate_nodes` method. The test would just make sure that we don't terminate more than 1000 nodes in a single request.

**Comment by hngenc** [comment]
@AmeerHajAli I just pushed a new unit test. It makes sure that the AWS Node Provider breaks large termination requests up into smaller chunks.

**Review by sasha-s — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ray-project/ray/17642)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @hngenc and @sasha-s)

___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 459 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr8S-kBczIKyF7ieac:-Mgr8S-kBczIKyF7iead:bx2697s) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L459)):*
> ```Python
>         pass
> 
>     def terminate_nodes(self, node_ids):
> ```

would it make sense to pull the max_terminate_nodes logic up, so it is shared across providers?
we might want to do something similar for GCP
this way, in each provider, we only need to specify max_terminate_nodes.
___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 485 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr7dkoBILNAtJKlj52:-Mgr7dkpEc_ea3QclZlS:bms5h6m) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L485)):*
> ```Python
>                     cli_logger.render_list(on_demand_ids))
> 
>                 for start in range(0, len(on_demand_ids), max_terminate_nodes):
> ```

maybe have a helper function that returns (ids to terminate, termination func)
for spot instances we would return (spot_ids,  self.ec2.meta.client.terminate_instances),
for on_demand_ids: (on_demand_ids, stop_instances)

and then have a single loop 
for start in range(0, len(ids),  max_terminate_nodes):
   terminate_func(ids[start:start +  max_terminate_nodes]

**Review by hngenc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ray-project/ray/17642)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @sasha-s)

___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 459 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr8S-kBczIKyF7ieac:-MgrFdHw8SFxS8G0PsFT:b-asf4d8) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L459)):*
<details><summary><i>Previously, sasha-s (Sasha Sobol) wrote…</i></summary><blockquote>

would it make sense to pull the max_terminate_nodes logic up, so it is shared across providers?
we might want to do something similar for GCP
this way, in each provider, we only need to specify max_terminate_nodes.
</blockquote></details>

Perhaps; although I have no idea if this limitation is shared across GCP and other providers. But I don't mind adding this if you think we should.

In that case, the default value of `max_terminate_nodes` would be `None`, and if it is None, then no loop would be used.
___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 485 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr7dkoBILNAtJKlj52:-MgrGONB3mSkYzYVd9SF:b8se4u) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L485)):*
<details><summary><i>Previously, sasha-s (Sasha Sobol) wrote…</i></summary><blockquote>

maybe have a helper function that returns (ids to terminate, termination func)
for spot instances we would return (spot_ids,  self.ec2.meta.client.terminate_instances),
for on_demand_ids: (on_demand_ids, stop_instances)

and then have a single loop 
for start in range(0, len(ids),  max_terminate_nodes):
   terminate_func(ids[start:start +  max_terminate_nodes]

</blockquote></details>

Instead of a helper func, what if we just set two variables `ids_to_terminate` and `terminate_func`, and set them in the `terminate_nodes()` function?

This would be identical to what you propose, but without  a new helper func, since I don't think the helper func would be used anywhere else other than in the `terminate_nodes` function, so I don't know how helpful it would be to pull out into another function.

**Review by sasha-s — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ray-project/ray/17642)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @hngenc and @sasha-s)

___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 459 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr8S-kBczIKyF7ieac:-MgrHBDo2VyLa0c5vh4D:bh78ob) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L459)):*
<details><summary><i>Previously, hngenc (Hasan Genc) wrote…</i></summary><blockquote>

Perhaps; although I have no idea if this limitation is shared across GCP and other providers. But I don't mind adding this if you think we should.

In that case, the default value of `max_terminate_nodes` would be `None`, and if it is None, then no loop would be used.
</blockquote></details>

sounds great!
___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 485 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr7dkoBILNAtJKlj52:-MgrHMPX7jmAV8qvil2y:bjhc6k2) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L485)):*
<details><summary><i>Previously, hngenc (Hasan Genc) wrote…</i></summary><blockquote>

Instead of a helper func, what if we just set two variables `ids_to_terminate` and `terminate_func`, and set them in the `terminate_nodes()` function?

This would be identical to what you propose, but without  a new helper func, since I don't think the helper func would be used anywhere else other than in the `terminate_nodes` function, so I don't know how helpful it would be to pull out into another function.
</blockquote></details>

also works

**Review by sasha-s — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ray-project/ray/17642)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @hngenc and @sasha-s)

___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 459 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr8S-kBczIKyF7ieac:-MgrHgegAT6uaBci6d1r:b-dajsx9) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L459)):*
<details><summary><i>Previously, sasha-s (Sasha Sobol) wrote…</i></summary><blockquote>

sounds great!
</blockquote></details>

alternatively you could always use the loop (setting max_terminate_nodes to len(ids) in case of None).
this would always exercise the same code path, which could be a (very minor) plus

**Review by hngenc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ray-project/ray/17642)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @hngenc and @sasha-s)

___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 459 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr8S-kBczIKyF7ieac:-MgrMW4RCdM2U90xiMvO:b6ab6r3) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L459)):*
<details><summary><i>Previously, sasha-s (Sasha Sobol) wrote…</i></summary><blockquote>

alternatively you could always use the loop (setting max_terminate_nodes to len(ids) in case of None).
this would always exercise the same code path, which could be a (very minor) plus 
</blockquote></details>

Sure that works

**Review by sasha-s — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ray-project/ray/17642)* status: all files reviewed, 1 unresolved discussion (waiting on @hngenc)

**Review by hngenc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ray-project/ray/17642)* status: all files reviewed, 1 unresolved discussion (waiting on @sasha-s)

___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 485 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr7dkoBILNAtJKlj52:-MgrNM2s2tQyG6n_l0Rr:b-4rrqvx) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L485)):*
<details><summary><i>Previously, sasha-s (Sasha Sobol) wrote…</i></summary><blockquote>

also works 
</blockquote></details>

Actually, it seems that we sometimes call both `terminate_instances`  and `stop_instances` within the same `terminate_nodes` call (such as if we have both spot and on-demand nodes). So we would need more than one `terminate_func`.

I'll think of an unconfusing way to implement this

**Review by sasha-s — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ray-project/ray/17642)* status: all files reviewed, 1 unresolved discussion (waiting on @hngenc)

___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 485 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr7dkoBILNAtJKlj52:-MgrNrNe-V6_lcrfJbSV:blc170e) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L485)):*
<details><summary><i>Previously, hngenc (Hasan Genc) wrote…</i></summary><blockquote>

Actually, it seems that we sometimes call both `terminate_instances`  and `stop_instances` within the same `terminate_nodes` call (such as if we have both spot and on-demand nodes). So we would need more than one `terminate_func`.

I'll think of an unconfusing way to implement this

</blockquote></details>

one possibility (maybe not the best one, but would probably work) is to compute a list of pairs ([ids, termination_func)] 
and then go over it and apply max_nodes_to_terminate

**Review by hngenc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ray-project/ray/17642)* status: all files reviewed, 1 unresolved discussion (waiting on @sasha-s)

___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 459 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr8S-kBczIKyF7ieac:-MgrcwP-3VgmjudftXJT:b5uuzdf) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L459)):*
<details><summary><i>Previously, hngenc (Hasan Genc) wrote…</i></summary><blockquote>

Sure that works
</blockquote></details>

I've added a `max_terminate_nodes` property to the `NodeProvider` class
___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 485 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr7dkoBILNAtJKlj52:-MgrcYOn31DAtQBoke3F:b-9l15ut) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L485)):*
<details><summary><i>Previously, sasha-s (Sasha Sobol) wrote…</i></summary><blockquote>

one possibility (maybe not the best one, but would probably work) is to compute a list of pairs ([ids, termination_func)] 
and then go over it and apply max_nodes_to_terminate

</blockquote></details>

I've implemented it like this:

```
nodes_to_terminate = {terminate_instances_func: [],
    stop_instances_func: []}
```

**Review by sasha-s — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ray-project/ray/17642#-:-MgssynT-UHysyxlwmon:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @hngenc)

___
*[python/ray/autoscaler/_private/aws/node_provider.py, line 485 at r3](https://reviewable.io/reviews/ray-project/ray/17642#-Mgr7dkoBILNAtJKlj52:-MgssUXo8DSGU-NhZoIS:bb70w76) ([raw file](https://github.com/ray-project/ray/blob/bb1f272d559f50df0e614ae7a55e4afd5975610e/python/ray/autoscaler/_private/aws/node_provider.py#L485)):*
<details><summary><i>Previously, hngenc (Hasan Genc) wrote…</i></summary><blockquote>

I've implemented it like this:

```
nodes_to_terminate = {terminate_instances_func: [],
    stop_instances_func: []}
```
</blockquote></details>

Nice

**Review by sasha-s — APPROVED** [review]
-(@author-action-required) 

> *[Reviewable](https://reviewable.io/reviews/ray-project/ray/17642#-:-MgstChc8C3Ms6F9xa9K:bri1gjv)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @hngenc)

File: python/ray/autoscaler/node_provider.py [hunk 1]
```diff
@@ -141,6 +141,22 @@ def terminate_nodes(self, node_ids: List[str]) -> Optional[Dict[str, Any]]:
...
+        with one request.
+
+        For example, AWS only allows 1000 nodes to be terminated
+        at once; to terminate more, we must issue multiple separate API
+        requests. If the limit is infinity, then simply set this to None.
+
+        This may be overridden. The value may be useful when overriding the
+        "terminate_nodes" method.
+        """
+        return None
```
Comment (AmeerHajAli) [inline python/ray/autoscaler/node_provider.py:158]: why do we need this? 
It does not seem to be needed now for the AWSNodeProvider. It does not seem to be doing anything anyway.

File: python/ray/tests/aws/test_autoscaler_aws.py [hunk 2]
```diff
@@ -622,6 +623,55 @@ def test_launch_templates(ec2_client_stub, ec2_client_stub_fail_fast,
...
+@pytest.mark.parametrize("stop", [True, False])
+@pytest.mark.parametrize("spot", [True, False])
+def test_terminate_nodes(num_nodes, stop, spot):
+    # This test makes sure that we don't try to stop or terminate too many
+    # nodes in a single EC2 request. By default, only 1000 nodes can be
+    # stopped/terminated in one request. To terminate more nodes, we must break
+    # them up into multiple smaller requests.
+    #
+    # "num_nodes" is the number of nodes to stop or terminate.
+    # "stop" is True if we want to stop nodes, and False otherwise. Note that
```
Comment (AmeerHajAli) [inline python/ray/tests/aws/test_autoscaler_aws.py:636]:
```suggestion
    # "stop" is True if we want to stop nodes, and False to terminate nodes. Note that
```

**Review by AmeerHajAli — APPROVED** [review]
LGTM. 2 minor comments. I think this is very close to be merged.

**Review by sasha-s — APPROVED** [review]
> Reviewed 2 of 2 files at r6, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ray-project/ray/17642)* status: all files reviewed, 2 unresolved discussions (waiting on @hngenc and @sasha-s)

**Review by AmeerHajAli — APPROVED** [review]

**Comment by AmeerHajAli** [comment]
Thanks @hngenc, this is a great usability fix!

File: python/ray/autoscaler/_private/aws/node_provider.py [hunk 3]
```diff
@@ -478,16 +495,25 @@ def terminate_nodes(self, node_ids):
...
+            nodes_to_terminate[terminate_instances_func] = spot_ids
         else:
-            self.ec2.meta.client.terminate_instances(InstanceIds=node_ids)
+            nodes_to_terminate[terminate_instances_func] = node_ids
+
+        max_terminate_nodes = self.max_terminate_nodes if \
+            self.max_terminate_nodes is not None else len(node_ids)
+
+        # for terminate_func, nodes in nodes_to_terminate.items():
+        for nodes, terminate_func in nodes_to_terminate.items():
```
Comment (iycheng) [inline python/ray/autoscaler/_private/aws/node_provider.py:513]: Should this be

```
for terminate_func, nodes in nodes_to_terminate.items():
```


I met errors like this:

```
Shared connection to 50.112.216.147 closed.
Traceback (most recent call last):
  File "/mnt/yic/.local/anaconda3/bin/ray", line 33, in <module>
    sys.exit(load_entry_point('ray', 'console_scripts', 'ray')())
  File "/mnt/yic/ray-w2/python/ray/scripts/scripts.py", line 1984, in main
    return cli()
  File "/mnt/yic/.local/anaconda3/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/mnt/yic/.local/anaconda3/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/mnt/yic/.local/anaconda3/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/mnt/yic/.local/anaconda3/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/mnt/yic/.local/anaconda3/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/mnt/yic/ray-w2/python/ray/scripts/scripts.py", line 1020, in down
    teardown_cluster(cluster_config_file, yes, workers_only, cluster_name,
  File "/mnt/yic/ray-w2/python/ray/autoscaler/_private/commands.py", line 449, in teardown_cluster
    provider.terminate_nodes(A)
  File "/mnt/yic/ray-w2/python/ray/autoscaler/_private/aws/node_provider.py", line 514, in terminate_nodes
    for start in range(0, len(nodes), max_terminate_nodes):
TypeError: object of type 'method' has no len()
```

Should we have `ray down` test to cover this code path?

Comment (hngenc) [reply python/ray/autoscaler/_private/aws/node_provider.py:513]: Sorry about that. I fixed it over here:

https://github.com/ray-project/ray/pull/17833

**Pull request merged**